src/main/frontend/fs/diff_merge.cljc

```diff
@@ -0,0 +1,93 @@
+(ns frontend.fs.diff-merge
+  ;; Disable clj linters since we don't support clj
+  #?(:clj {:clj-kondo/config {:linters {:unresolved-namespace {:level :off}
+                                        :unresolved-symbol {:level :off}}}})
+  (:require #?(:org.babashka/nbb ["@logseq/diff-merge$default" :refer [Merger Differ visualizeAsHTML attach_uuids]]
+               :default ["@logseq/diff-merge" :refer [Differ Merger visualizeAsHTML attach_uuids]])
+            [logseq.graph-parser.block :as gp-block]
+            [logseq.graph-parser.property :as gp-property]
+            [logseq.graph-parser.utf8 :as utf8]
+            [cljs-bean.core :as bean]
+            [frontend.db.utils :as db-utils]
+            [frontend.db.model :as db-model]))
+
+;; (defn diff-merge
+;;   "N-ways diff & merge
+;;    Accept: blocks
+;;    https://github.com/logseq/diff-merge/blob/44546f2427f20bd417b898c8ba7b7d10a9254774/lib/mldoc.ts#L17-L22
+;;    https://github.com/logseq/diff-merge/blob/85ca7e9bf7740d3880ed97d535a4f782a963395d/lib/merge.ts#L40"
+;;   [base & branches]
+;;   ()
+;;   (let [merger (Merger.)]
+;;     (.mergeBlocks merger (bean/->js base) (bean/->js branches))))
+
+(defn diff 
+  "2-ways diff
+   Accept: blocks
+   https://github.com/logseq/diff-merge/blob/44546f2427f20bd417b898c8ba7b7d10a9254774/lib/mldoc.ts#L17-L22"
+  [base income]
+  (let [differ (Differ.)]
+    (.diff_logseqMode differ (bean/->js base) (bean/->js income))))
+
+;; (defonce getHTML visualizeAsHTML)
+
+(defonce attachUUID attach_uuids)
+
+(defn db->diff-blocks
+  "db: datascript db
+   page-name: string"
+  [page-name]
+  {:pre (string? page-name)}
+  (let [walked (db-model/get-sorted-page-block-ids-and-levels page-name)
+        blocks (db-utils/pull-many [:block/uuid :block/content :block/level] (map :id walked))
+        levels (map :level walked)
+        blocks (map (fn [block level]
+                      {:uuid   (str (:block/uuid block)) ;; Force to be string
+                       :body   (:block/content block)
+                       :level  level})
+                    blocks levels)]
+    blocks))
+
+;; TODO Junyi: merge back to gp-block/extract-blocks
```

Comment:
If we change the implementation to gp-block/extract-blocks, does this functionality break? If so, we should do the merge back in this PR so there isn't is the coupling between these two fns. Happy to help if you could describe how to QA this fn (or I could rely on tests if that is enough)

Reply:
Hum, it's nearly impossible to merge back as I want to decouple diff-merge with graph parser. So I'm removing the TODO here :(

> Happy to help if you could describe how to QA this fn

The current process of doing 2 way diff is like:
1. Given a base ver. of page (AST in DB), and a branch ver. of page (externally modified file content)
2. Transform both base ver. & branch ver. into the same format (`diff-blocks`)
3. Apply `diff-merge/diff` on them, which returns the resolved uuids (the uuids of the corresponding blocks in DB) of the branch ver. (externally modified file content)
4. Attach these resolved uuids into the blocks newly parsed by graph-parser

What I'm doing in the PR is to keep all the `diff-merge` fns, including `diff-merge/ast->diff-blocks`, out of the `graph-parser`, and only inject the step 4 into `graph-parser` as a hook
Code: 
https://github.com/logseq/logseq/blob/edc87dc0fe55a26d6232ebf5d530d08132193560/src/main/frontend/handler/common/file.cljs#L57

For how to QA:
In previous version of Logseq, any external editing to a `.md` / `.org` file should "refresh" all blocks' uuids on that page, except those with `id::` provided. 
In this PR, the external editing will not "refresh" the blocks' uuids, except the block that was edited externally.
There's a test for checking if the block uuids are persisted upon file parsing:
https://github.com/logseq/logseq/blob/3c0175dfc29899da45e65781f6c88aefbf84a80b/src/test/frontend/fs/diff_merge_test.cljs#L307-L340

Reply:
> merge back as I want to decouple diff-merge with graph parser.

Decoupling the two sounds like a good idea but there is still a fair amount of indirect coupling of ast knowledge. Hopefully this doesn't break the next time `gp-block/extract-blocks` changes